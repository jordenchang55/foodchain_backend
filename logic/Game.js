import { shuffle } from '../utils/arr';
import { mapTiles } from '../gameConfigs';
import GameMap from './GameMap';
import { transformPosition } from '../utils/map';
import Pool from './Pool';
import Structure from './Structure';

export const setupMap = (playerNumber) => {
    const xSize = playerNumber >= 3 ? playerNumber : 3;
    const ySize = playerNumber >= 3 ? 4 : 3;
    const tileIds = shuffle(Object.keys(mapTiles)).slice(0, xSize * ySize);
    const configs = tileIds
        .map((tileId) => ({
            tileId,
            direction: Math.floor(Math.random() * 4),
        }));
    return new GameMap(xSize, ySize, configs);
};

export const generateWorkingOrder = (players) => {
    const order = [];
    for (let i = 0; i < players.length; i += 1) {
        order.push(i);
    }
    const randomOrder = shuffle(order);
    return randomOrder.map((i) => players[i]);
};

export default class Game {
    constructor(eventManager) {
        this.eventManager = eventManager;
    }

    initialize(players) {
        this.turn = 1;
        this.phase = 1;
        this.playerProps = players.reduce((map, playerName) => ({
            ...map,
            [playerName]: {
                coffeeShopAmount: 3,
                restaurantAmount: 3,
                foods: [],
                milestones: [],
                structure: undefined,
                onBeach: {},
                advertising: {},
                money: 0,
            },
        }), {});
        this.reserveCards = {};
        this.lastWorkingOrder = {
            // Username: index
        };
        this.restaurants = {};
    }

    setup() {
        const players = Object.keys(this.playerProps);
        this.gameMap = setupMap(players.length);
        this.eventManager.notifyAll('setup_map', {
            tiles: this.gameMap.tileConfig,
        });

        this.pool = new Pool(players.length);
        this.eventManager.notifyAll('pool_update', this.pool.toObject());

        this.workingOrder = generateWorkingOrder(players);
        this.eventManager.notifyAll('order_decision', {
            available: [],
            selected: Object.keys(this.workingOrder).sort((a, b) => this.workingOrder[a] - this.workingOrder[b]),
        });

        this.setupRestaurants()
            .then(() => this.askReserveCards())
            .then(() => {
                this.eventManager.on(
                    'lock_structure',
                    (username, msg) => this.onStructureLocked(username, msg, this.askOrderDecision),
                );
                this.eventManager.on('pick_order', this.onOrderDecided);
                this.eventManager.notifyAll('turn_update', {
                    turn: this.turn,
                });
            });
    }

    setupRestaurants() {
        return new Promise((resolve) => {
            this.eventManager.on(
                'pick_first_restaurant',
                (username, msg) => this.onFirstRestaurantPick(username, [
                    msg.position.xTile,
                    msg.position.yTile,
                    msg.position.xSmall,
                    msg.position.ySmall,
                ], msg.direction, resolve),
            );
            const firstPick = this.workingOrder[this.workingOrder.length - 1];
            this.askFirstRestaurant(firstPick, true);
        });
    }

    askFirstRestaurant(username, skippable) {
        this.eventManager.notifyAll('first_restaurant_decision', {
            nextPlayer: username,
            skippable,
        });
    }

    onFirstRestaurantPick(username, position, direction, done) {
        if (!position) {
            this.restaurants[username] = null;
        } else {
            this.gameMap.putTile(username, position.map((p) => transformPosition(p, direction)));
            this.restaurants[username] = 1;
            this.playerProps[username].restaurantAmount = 2;
            this.eventManager.notifyAll('tiles_placement', {
                type: 'restaurant',
                position,
                direction,
            });
            this.eventManager.notifyAll('player_update', this.playerProps);
        }
        let nextIndex = (this.workingOrder.indexOf(username) - 1 + this.workingOrder.length);

        while (nextIndex >= 0 && this.restaurants[this.workingOrder[nextIndex % this.workingOrder.length]]) {
            nextIndex -= 1;
        }
        if (nextIndex < 0) {
            done?.();
        } else {
            const nextUser = this.workingOrder[nextIndex % this.workingOrder.length];
            this.askFirstRestaurant(nextUser, this.restaurants[nextUser] === undefined);
        }
    }

    askReserveCards() {
        return new Promise((resolve) => {
            this.eventManager.on('pick_reserve_card', (username, msg) => {
                this.onReserveCardPick(username, msg.card, resolve);
            });
            this.eventManager.notifyAll('reserve_card_decision', {});
        });
    }

    onReserveCardPick(username, card, done) {
        this.reserveCards[username] = card;
        if (Object.keys(this.reserveCards).length === Object.keys(this.playerProps).length) {
            done?.();
        }
    }

    /**
     * When a player finalize their structure, this callback will be invoked.
     * @param username key of that player.
     * @param structure a map of employee id with number of that card.
     */
    onStructureLocked(username, structure, onComplete) {
        this.playerProps[username].structure = new Structure(structure);
        const emptyStructures = Object.values(this.playerProps)
            .map((props) => props.structure)
            .filter((s) => !s);
        if (!emptyStructures.length) {
            const structures = Object.keys(this.playerProps)
                .reduce((obj, player) => ({
                    ...obj,
                    [player]: {
                        structure: this.playerProps[player].structure.structure,
                        // TODO add milestone and reserve card offset
                        openSlots: this.playerProps[player].structure?.getOpenSlot() ?? 0,
                    },
                }), {});
            this.eventManager.notifyAll('show_up', structures);

            this.decisionOrder = Object.keys(this.playerProps)
                .sort((a, b) => {
                    if (structures[b].openSlots !== structures[a].openSlots) {
                        return structures[b].openSlots - structures[a].openSlots;
                    }
                    if (structures[b].structure.E01 !== structures[a].structure.E01) {
                        // Compare the number of waitresses if tie
                        return (structures[b].structure.E01 ?? 0) - (structures[a].structure.E01 ?? 0);
                    }
                    // Use the working order of last turn for tie breaker
                    return this.lastWorkingOrder[a] - this.lastWorkingOrder[b];
                });
            onComplete?.();
        }
    }
}
