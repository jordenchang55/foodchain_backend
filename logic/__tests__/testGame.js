import Game, { generateWorkingOrder, setupMap } from '../Game';
import EventManager from '../../models/EventManager';

jest.mock('../../models/EventManager');
expect.extend({
    toBeDistinct(received) {
        const pass = Array.isArray(received) && new Set(received).size === received.length;
        if (pass) {
            return {
                message: () => `expected [${received}] array is unique`,
                pass: true,
            };
        }
        return {
            message: () => `expected [${received}] array is not to unique`,
            pass: false,
        };
    },
});

describe('test Game', () => {
    let eventMgr;
    beforeEach(() => {
        EventManager.mockClear();
        eventMgr = new EventManager();
        jest.spyOn(global.Math, 'random').mockReturnValue(0.1);
    });
    describe('generateWorkingOrder', () => {
        it('success', () => {
            const order = generateWorkingOrder(['user1', 'user2', 'user3']);
            expect(order).toMatchSnapshot();
        });
    });

    describe('askFirstRestaurant', () => {
        it('not skippable', () => {
            const game = new Game(eventMgr);
            game.askFirstRestaurant('user1', false);
            const mockEventMgr = EventManager.mock.instances[0];
            expect(mockEventMgr.notifyAll).toBeCalledWith('first_restaurant_decision', {
                nextPlayer: 'user1',
                skippable: false,
            });
        });
        it('skippable', () => {
            const game = new Game(eventMgr);
            game.askFirstRestaurant('user1', true);
            const mockEventMgr = EventManager.mock.instances[0];
            expect(mockEventMgr.notifyAll).toBeCalledWith('first_restaurant_decision', {
                nextPlayer: 'user1',
                skippable: true,
            });
        });
    });

    describe('setupMap', () => {
        it('2 players', () => {
            const gameMap = setupMap(2);
            expect(gameMap.tileConfig.length).toBe(9);
            expect(gameMap.tileConfig.map((t) => t.tileId)).toBeDistinct();
            expect(gameMap.tileConfig.map((t) => t.position).map((t) => JSON.stringify(t)))
                .toBeDistinct();
        });
        it('3 players', () => {
            const gameMap = setupMap(3);
            expect(gameMap.tileConfig.length).toBe(12);
            expect(gameMap.tileConfig.map((t) => t.tileId)).toBeDistinct();
            expect(gameMap.tileConfig.map((t) => t.position).map((t) => JSON.stringify(t)))
                .toBeDistinct();
        });
        it('4 players', () => {
            const gameMap = setupMap(4);
            expect(gameMap.tileConfig.length).toBe(16);
            expect(gameMap.tileConfig.map((t) => t.tileId)).toBeDistinct();
            expect(gameMap.tileConfig.map((t) => t.position).map((t) => JSON.stringify(t)))
                .toBeDistinct();
        });
        it('5 players', () => {
            const gameMap = setupMap(5);
            expect(gameMap.tileConfig.length).toBe(20);
            expect(gameMap.tileConfig.map((t) => t.tileId)).toBeDistinct();
            expect(gameMap.tileConfig.map((t) => t.position).map((t) => JSON.stringify(t)))
                .toBeDistinct();
        });
    });

    describe('initialize', () => {
        it('normal', () => {
            const game = new Game(eventMgr);
            game.initialize(['user1', 'user2', 'user3']);
            game.setup();
            const mockEventMgr = EventManager.mock.instances[0];
            const done = jest.fn();
            game.onFirstRestaurantPick('user1', [
                [0, 0, 3, 3],
                [0, 0, 4, 3],
                [0, 0, 3, 4],
                [0, 0, 4, 4],
            ], 0, done);
            game.onFirstRestaurantPick('user3', [
                [2, 3, 3, 3],
                [2, 3, 4, 3],
                [2, 3, 3, 4],
                [2, 3, 4, 4],
            ], 0, done);
            game.onFirstRestaurantPick('user2', [
                [0, 3, 3, 3],
                [0, 3, 4, 3],
                [0, 3, 3, 4],
                [0, 3, 4, 4],
            ], 0, done);
            expect(done).toBeCalledTimes(1);
            expect(mockEventMgr.notifyAll.mock.calls).toMatchSnapshot();
        });
        it('skip restaurant pick', () => {
            const game = new Game(eventMgr);
            game.initialize(['user1', 'user2', 'user3']);
            game.setup();
            const mockEventMgr = EventManager.mock.instances[0];
            game.onFirstRestaurantPick('user1', undefined);
            game.onFirstRestaurantPick('user3', undefined);
            game.onFirstRestaurantPick('user2', undefined);
            expect(mockEventMgr.notifyAll.mock.calls).toMatchSnapshot();
        });
    });

    describe('askReserveCards', () => {
        it('normal', () => {
            const game = new Game(eventMgr);
            game.initialize(['user1', 'user2', 'user3']);
            const mockEventMgr = EventManager.mock.instances[0];
            game.askReserveCards();
            expect(mockEventMgr.notifyAll.mock.calls[0][0]).toBe('reserve_card_decision');
            const done = jest.fn();
            game.onReserveCardPick('user1', 100, done);
            game.onReserveCardPick('user2', 200, done);
            game.onReserveCardPick('user3', 300, done);
            expect(done).toBeCalledTimes(1);
        });
    });
});
