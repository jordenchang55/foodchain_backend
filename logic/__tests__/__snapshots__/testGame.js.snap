// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Game generateWorkingOrder success 1`] = `
Object {
  "user1": 3,
  "user2": 1,
  "user3": 2,
}
`;

exports[`test Game initialize normal 1`] = `
Array [
  Array [
    "setup_map",
    Object {
      "tiles": Array [
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 0,
          },
          "tileId": "M10",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 0,
          },
          "tileId": "M20",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 0,
          },
          "tileId": "M04",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 1,
          },
          "tileId": "M05",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 1,
          },
          "tileId": "M06",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 1,
          },
          "tileId": "M07",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 2,
          },
          "tileId": "M08",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 2,
          },
          "tileId": "M09",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 2,
          },
          "tileId": "M01",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 3,
          },
          "tileId": "M11",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 3,
          },
          "tileId": "M12",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 3,
          },
          "tileId": "M13",
        },
      ],
    },
  ],
  Array [
    "pool_update",
    Object {
      "ads": Array [
        "Mk01",
        "Mk02",
        "Mk03",
        "Mk04",
        "Mk05",
        "Mk06",
        "Mk07",
        "Mk08",
        "Mk09",
        "Mk10",
        "Mk11",
        "Mk12",
        "Mk13",
        "Mk14",
      ],
      "employees": Object {
        "E01": 12,
        "E02": 18,
        "E03": 12,
        "E04": 12,
        "E05": 12,
        "E06": 12,
        "E07": 12,
        "E08": 12,
        "E09": 6,
        "E10": 12,
        "E11": 1,
        "E12": 6,
        "E13": 6,
        "E14": 6,
        "E15": 6,
        "E16": 6,
        "E17": 6,
        "E18": 6,
        "E19": 6,
        "E20": 6,
        "E21": 6,
        "E22": 6,
        "E23": 1,
        "E24": 1,
        "E25": 1,
        "E26": 6,
        "E27": 3,
        "E28": 1,
        "E29": 1,
        "E30": 1,
        "E31": 1,
        "E32": 1,
      },
      "gardenNumber": 8,
      "houses": Array [
        "H01",
        "H03",
        "H06",
        "H09",
        "H11",
        "H14",
        "H17",
        "H19",
      ],
      "milestones": Object {
        "achieved": Object {},
        "new": Object {},
        "remain": Array [
          "Mi01",
          "Mi02",
          "Mi03",
          "Mi04",
          "Mi05",
          "Mi06",
          "Mi07",
          "Mi08",
          "Mi09",
          "Mi10",
          "Mi11",
          "Mi12",
          "Mi13",
          "Mi14",
          "Mi15",
          "Mi16",
          "Mi17",
          "Mi18",
        ],
      },
      "money": 150,
    },
  ],
  Array [
    "order_decision",
    Object {
      "available": Array [],
      "selected": Array [
        "user2",
        "user3",
        "user1",
      ],
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user1",
      "skippable": true,
    },
  ],
  Array [
    "tiles_placement",
    Object {
      "direction": 0,
      "position": Array [
        Array [
          0,
          0,
          3,
          3,
        ],
        Array [
          0,
          0,
          4,
          3,
        ],
        Array [
          0,
          0,
          3,
          4,
        ],
        Array [
          0,
          0,
          4,
          4,
        ],
      ],
      "type": "restaurant",
    },
  ],
  Array [
    "player_update",
    Object {
      "user1": Object {
        "advertising": Object {},
        "coffeeShopAmount": 3,
        "foods": Array [],
        "milestones": Array [],
        "money": 0,
        "onBeach": Object {},
        "restaurantAmount": 2,
        "structure": Object {},
      },
      "user2": Object {
        "advertising": Object {},
        "coffeeShopAmount": 3,
        "foods": Array [],
        "milestones": Array [],
        "money": 0,
        "onBeach": Object {},
        "restaurantAmount": 2,
        "structure": Object {},
      },
      "user3": Object {
        "advertising": Object {},
        "coffeeShopAmount": 3,
        "foods": Array [],
        "milestones": Array [],
        "money": 0,
        "onBeach": Object {},
        "restaurantAmount": 2,
        "structure": Object {},
      },
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user3",
      "skippable": true,
    },
  ],
  Array [
    "tiles_placement",
    Object {
      "direction": 0,
      "position": Array [
        Array [
          2,
          3,
          3,
          3,
        ],
        Array [
          2,
          3,
          4,
          3,
        ],
        Array [
          2,
          3,
          3,
          4,
        ],
        Array [
          2,
          3,
          4,
          4,
        ],
      ],
      "type": "restaurant",
    },
  ],
  Array [
    "player_update",
    Object {
      "user1": Object {
        "advertising": Object {},
        "coffeeShopAmount": 3,
        "foods": Array [],
        "milestones": Array [],
        "money": 0,
        "onBeach": Object {},
        "restaurantAmount": 2,
        "structure": Object {},
      },
      "user2": Object {
        "advertising": Object {},
        "coffeeShopAmount": 3,
        "foods": Array [],
        "milestones": Array [],
        "money": 0,
        "onBeach": Object {},
        "restaurantAmount": 2,
        "structure": Object {},
      },
      "user3": Object {
        "advertising": Object {},
        "coffeeShopAmount": 3,
        "foods": Array [],
        "milestones": Array [],
        "money": 0,
        "onBeach": Object {},
        "restaurantAmount": 2,
        "structure": Object {},
      },
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user2",
      "skippable": true,
    },
  ],
  Array [
    "tiles_placement",
    Object {
      "direction": 0,
      "position": Array [
        Array [
          0,
          3,
          3,
          3,
        ],
        Array [
          0,
          3,
          4,
          3,
        ],
        Array [
          0,
          3,
          3,
          4,
        ],
        Array [
          0,
          3,
          4,
          4,
        ],
      ],
      "type": "restaurant",
    },
  ],
  Array [
    "player_update",
    Object {
      "user1": Object {
        "advertising": Object {},
        "coffeeShopAmount": 3,
        "foods": Array [],
        "milestones": Array [],
        "money": 0,
        "onBeach": Object {},
        "restaurantAmount": 2,
        "structure": Object {},
      },
      "user2": Object {
        "advertising": Object {},
        "coffeeShopAmount": 3,
        "foods": Array [],
        "milestones": Array [],
        "money": 0,
        "onBeach": Object {},
        "restaurantAmount": 2,
        "structure": Object {},
      },
      "user3": Object {
        "advertising": Object {},
        "coffeeShopAmount": 3,
        "foods": Array [],
        "milestones": Array [],
        "money": 0,
        "onBeach": Object {},
        "restaurantAmount": 2,
        "structure": Object {},
      },
    },
  ],
]
`;

exports[`test Game initialize skip restaurant pick 1`] = `
Array [
  Array [
    "setup_map",
    Object {
      "tiles": Array [
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 0,
          },
          "tileId": "M10",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 0,
          },
          "tileId": "M20",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 0,
          },
          "tileId": "M04",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 1,
          },
          "tileId": "M05",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 1,
          },
          "tileId": "M06",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 1,
          },
          "tileId": "M07",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 2,
          },
          "tileId": "M08",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 2,
          },
          "tileId": "M09",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 2,
          },
          "tileId": "M01",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 0,
            "yTile": 3,
          },
          "tileId": "M11",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 1,
            "yTile": 3,
          },
          "tileId": "M12",
        },
        Object {
          "direction": 0,
          "position": Object {
            "xTile": 2,
            "yTile": 3,
          },
          "tileId": "M13",
        },
      ],
    },
  ],
  Array [
    "pool_update",
    Object {
      "ads": Array [
        "Mk01",
        "Mk02",
        "Mk03",
        "Mk04",
        "Mk05",
        "Mk06",
        "Mk07",
        "Mk08",
        "Mk09",
        "Mk10",
        "Mk11",
        "Mk12",
        "Mk13",
        "Mk14",
      ],
      "employees": Object {
        "E01": 12,
        "E02": 18,
        "E03": 12,
        "E04": 12,
        "E05": 12,
        "E06": 12,
        "E07": 12,
        "E08": 12,
        "E09": 6,
        "E10": 12,
        "E11": 1,
        "E12": 6,
        "E13": 6,
        "E14": 6,
        "E15": 6,
        "E16": 6,
        "E17": 6,
        "E18": 6,
        "E19": 6,
        "E20": 6,
        "E21": 6,
        "E22": 6,
        "E23": 1,
        "E24": 1,
        "E25": 1,
        "E26": 6,
        "E27": 3,
        "E28": 1,
        "E29": 1,
        "E30": 1,
        "E31": 1,
        "E32": 1,
      },
      "gardenNumber": 8,
      "houses": Array [
        "H01",
        "H03",
        "H06",
        "H09",
        "H11",
        "H14",
        "H17",
        "H19",
      ],
      "milestones": Object {
        "achieved": Object {},
        "new": Object {},
        "remain": Array [
          "Mi01",
          "Mi02",
          "Mi03",
          "Mi04",
          "Mi05",
          "Mi06",
          "Mi07",
          "Mi08",
          "Mi09",
          "Mi10",
          "Mi11",
          "Mi12",
          "Mi13",
          "Mi14",
          "Mi15",
          "Mi16",
          "Mi17",
          "Mi18",
        ],
      },
      "money": 150,
    },
  ],
  Array [
    "order_decision",
    Object {
      "available": Array [],
      "selected": Array [
        "user2",
        "user3",
        "user1",
      ],
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user1",
      "skippable": true,
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user3",
      "skippable": true,
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user2",
      "skippable": true,
    },
  ],
  Array [
    "first_restaurant_decision",
    Object {
      "nextPlayer": "user1",
      "skippable": false,
    },
  ],
]
`;

exports[`test Game lock structure empty structure 1`] = `
Array [
  Array [
    "show_up",
    Object {
      "user1": Object {
        "openSlots": 3,
        "structure": Object {},
      },
      "user2": Object {
        "openSlots": 3,
        "structure": Object {},
      },
      "user3": Object {
        "openSlots": 3,
        "structure": Object {},
      },
    },
  ],
]
`;

exports[`test Game lock structure one employee structure 1`] = `
Array [
  Array [
    "show_up",
    Object {
      "user1": Object {
        "openSlots": 2,
        "structure": Object {
          "E01": 1,
        },
      },
      "user2": Object {
        "openSlots": 4,
        "structure": Object {
          "E02": 1,
        },
      },
      "user3": Object {
        "openSlots": 2,
        "structure": Object {
          "E03": 1,
        },
      },
    },
  ],
]
`;
